### ISSUE TITLE
Long deleting records

### ISSUE THREAD

**Issue by crysan** [realm/realm-java#1462]
Hi!

There is a basis Realm Java 0.82.2.
Table Author - 1 record
Table Books - 46 entries associated with one entry Author

Code:
RealmResults<Authors> author = realm.where(Authors.class).equalTo("url_base", url_base).findAll();
RealmResults<Books> books = realm.where(Books.class).equalTo("author.url_base", url_base).findAll();

realm.beginTransaction();
books.clear();
author.clear();
realm.commitTransaction();

Deleting records time is from 2 to 8 minutes.
With what it can be connected?
Thank you!

**Comment by kneth** [comment]
That's sound like far too long time. Do you do any write transactions in other threads at the same time? I mean, `beginTransaction()` is blocking so only one write transaction can modify your Realm.

**Comment by crysan** [comment]
Nothing is done.  Just waiting for the record to be deleted.

**Comment by kneth** [comment]
@crysan Can you share your code? If only privately, send an email to help@realm.io.

**Comment by crysan** [comment]
public void deleteAuthor(final String url_base){
        new MaterialDialog.Builder(getActivity())
                .title(R.string.dialog_author_delete_title)
                .positiveText(R.string.dialog_author_delete_positive)
                .negativeText(R.string.dialog_author_delete_negative)
                .callback(new MaterialDialog.ButtonCallback() {
                    @Override
                    public void onPositive(MaterialDialog dialog) {

                        RealmResults<Authors> author = realm.where(Authors.class).equalTo("url_base", url_base).findAll();
                        RealmResults<Books> books = realm.where(Books.class).equalTo("author.url_base", url_base).findAll();

                        realm.beginTransaction();
                            books.clear();
                            author.clear();
                        realm.commitTransaction();

                        reloadAdapter();
                    }
                })
                .show();
    }

**Comment by kneth** [comment]
It is indeed a very simple case, and it shouldn't take that long. I will label it as a bug and we will investigate it further.

**Comment by crysan** [comment]
Good.  I will wait for the result.  Thank you!

**Comment by Anatidae-Project** [comment]
Wondering if there has been any movement on this issue. I have a database that's basically pulling in news feeds and storing the data in Realm. That works great. Love Realm's speed and ease.

But when I need to delete a bunch of records (say a few thousand) Realm can hang for 10-20 seconds easy. And if I have UI elements waiting for realm, that can block the UI experience (even if this is on a background thread). Blocking in this case meaning the user might be stuck with a progress bar for what seems like a VERY long time to read data as it is locked for writes. 

I'd love to see either way faster removing of records or some cool way to do it non-blocking of reads.

**Comment by kneth** [comment]
@Anatidae-Project Is it possible for you to share the code? Does your model contain relationships as @crysan's model? And how do you delete the objects (one-by-one in small transactions or using `RealmResults.clear()`?

**Comment by cmelchior** [comment]
Hi @crysan and @Anatidae-Project I tried to reproduce your cases and so far has come up short. From what I can tell deletes are fast, even with large amounts of data. Can you reproduce the error in a minimal test case project?

**Comment by cmelchior** [comment]
@Anatidae-Project Also note that in Realm (unlike SQLite) writes doesn't block reads. So you should be able to display data on the screen while data is being deleted in the background.

**Comment by kneth** [comment]
@crysan Could you add `System.nanoTimer()` or `Debug.threadCpuTimer()` before and after your transaction?

**Comment by crysan** [comment]
Android Studio 1.4R2 said functions nanoTimer () and threadCpuTimer () does not.

**Comment by crysan** [comment]
Found function System.nanoTime ();  but is not the case.  Delete records as long.  But!  I delete the database and re-create it.  Completed set of records ( approximately 10,000 with connections to 300 ) and had no problems.  All it removed very quickly.  Perhaps that was the reason it was in the old database file.  While other options I have.  I will be testing on.  Thank you for understanding.

**Comment by Anatidae-Project** [comment]
I'll put some timers in it tomorrow.
So, the model is an Article. It contains a number of primitive data types and an RealmList of ContentModel objects. These content model objects could contain a single instance of another ContentModel side.

Basically, the article is the main news. The content model contains either a value filling out text, image, or video data. Of course an article is made up of an array of paragraphs, images, etc...

I'll typically have 2k articles and around 28k contents in the database.

When I update the articles I keep track of what articles have been orphaned. Then I loop through these and delete if they are valid to do so. Here is the code:

```
    public static void clearArticle(Context context, List<String> ids) {
        Logger.d("Opening clearRealm instance");
        Realm clearRealm = Realm.getInstance(new RealmConfiguration.Builder(context).name("feeds").build());
        clearRealm.refresh();
        List<ArticleModel> results = new ArrayList<>();
        List<RealmResults<ContentModel>> contents = new ArrayList<>();
        int contentsCounter = 0;
        // Finds everything that might match.
        for (int i = 0; i < ids.size(); i++) {
            if (ids.get(i).equals( AppNYDN.getInstance().getSharedPreferences().getString(Constants.NOTIFICATON_KEY, "") )) {
                continue; // Skip over this instance
            }
            ArticleModel article = clearRealm.where(ArticleModel.class).equalTo("contentId", ids.get(i)).notEqualTo("bookmarked", true).findFirst();
            if (article != null) {
                results.add(article);
                RealmResults<ContentModel> foundContents = clearRealm.where(ContentModel.class).equalTo("parentId", ids.get(i)).notEqualTo("bookmarked", true).findAll();
                if (foundContents != null && foundContents.size() > 0) {
                    contents.add(foundContents);
                    contentsCounter += foundContents.size();
                }
            }
        }
        // Now clear them out.
        Logger.d("Removing data from Realm.\n\nArticles: " + results.size() + "\nContents: " + contentsCounter);
        for (int ic = 0; ic < contents.size(); ic++) {
            clearRealm.beginTransaction();
            contents.get(ic).clear();
            clearRealm.commitTransaction();
            clearRealm.refresh();

        }
        for (int ir = 0; ir < results.size(); ir++) {
            if (results.get(ir).isValid()) {
                clearRealm.beginTransaction();
                results.get(ir).removeFromRealm();
                clearRealm.commitTransaction();
                clearRealm.refresh();
            }
        }
        if (clearRealm != null) {
            clearRealm.close();
            Logger.d("Closing clearRealm instance");
        } else {
            Logger.w("Realm already closed... that shouldn't be");
        }
    }
```

When I did the above sans-refresh, I would often get OOM errors during this operation. During the delete, my whole app will sometimes freeze. Although, this is all happening on a background thread.

**Comment by kneth** [comment]
The OOM errors might come from the fact that you are creating a lot of temporary objects. Instead of `contents.add(foundContents)`, you could do `foundContents.clear()`. Likewise, you could build up a `RealmQuery<ArticleModel>` using the `ids` and logical-or (`or()`) and `clear` the `RealmResults<ArticleModel`. It would safe you some memory and loops.

If your app freeze, I guess that you have other write transactions at the same time.

**Comment by Anatidae-Project** [comment]
I simplifed to this:

```
    public static void clearArticle(Context context, List<String> ids) {
        Logger.d("Opening clearRealm instance");
        Realm clearRealm = Realm.getInstance(new RealmConfiguration.Builder(context).name("feeds").build());
        clearRealm.refresh();
        List<ArticleModel> results = new ArrayList<>();
        List<RealmResults<ContentModel>> contents = new ArrayList<>();
        int contentsCounter = 0;
        // Finds everything that might match.
        for (int i = 0; i < ids.size(); i++) {
            if (ids.get(i).equals( AppNYDN.getInstance().getSharedPreferences().getString(Constants.NOTIFICATON_KEY, "") )) {
                continue; // Skip over this instance
            }
            clearRealm.beginTransaction();

            try {
                clearRealm.where(ArticleModel.class).equalTo("contentId", ids.get(i)).notEqualTo("bookmarked", true).findAll().clear();
            } catch (Exception e) {
                // Should only throw if something is wrong with Realm.
            }
            try {
                clearRealm.where(ContentModel.class).equalTo("parentId", ids.get(i)).notEqualTo("bookmarked", true).findAll().clear();
            } catch (Exception e) {
                // Should only throw if something is wrong with Realm.
            }

            clearRealm.commitTransaction();
}
```

Clearing 307 articles and likely 3000 or so Contents. These are my two timestamps logged, the opening and closing of ClearRealm instance.
07:22:30.535 Opening
07:23:28.105 Closing

There was no other writing happening during this event. There was likely reading in the UI thread as articles are displayed on my first fragment. That fragment reads about 40 articles to fill out a few variables. 

I don't know if this makes a difference, but all my data coming in is in JSON format and I let Realm parse it. Sometimes it's new data that exists an existing article entry. Almost a minute to find and delete those records seems tooi long.

**Comment by Anatidae-Project** [comment]
I tried taking out the .refresh() call. Now, I only had 5 articles to remove this refresh.

07:40:28.671  start
07:40:30.471  end

That's more of the timing I'd like to see, but with all 307 records I cleared earlier.

**Comment by Anatidae-Project** [comment]
Deleting 318 ArticleModels and 28,623 ContentModels

Start sequence above:  01:05:23.368
End sequence above: 01:06:13.588
Time: aprox 50sec.

Is that what would be expected?

**Comment by cmelchior** [comment]
What device are you testing on?

**Comment by kneth** [comment]
@Anatidae-Project Can you share the entire app with us? If you can't do it in public, you can send a private email to help@realm.io.

**Comment by Anatidae-Project** [comment]
Yes, I'll get it sent by tomorrow AM to look at. @cmelchior I'm using a Note 4 and Nexus 6.

**Comment by emanuelez** [comment]
We didn't receive your project yet. Any news on this?

**Comment by Zhuinden** [comment]
Hello, I'm deleting a large amount of data on a background thread, but querying the table in the meantime freezes the UI.

**Comment by cmelchior** [comment]
Hi @Zhuinden Thank you. Are you able to reproduce this in a small sample project?

**Comment by kneth** [comment]
We haven't seen any this kind of poor performance in our internal benchmarks. @Zhuinden @Brian-McBride @crysan Can you provide us with an example so we can do some measurements? And enhance the performance, of course.

**Comment by Zhuinden** [comment]
@kneth I've been flooded with bug fixes and exhaustion so I haven't had the time to create this specific case yet in my realm mini-project where I'm loading and deleting a bunch of stuff (gotta add a query on top), but I'll get to it when I can.

**Comment by kneth** [comment]
@Zhuinden No worries, we all know what a bug flood feels like :cry:

**Comment by Zhuinden** [comment]
Okay, while I haven't had the opportunity to make a benchmark, I've managed to optimize out the issue that was causing ANRs on the main thread.

What happened is the following,

        @Override
        public void onRun()
                throws Throwable {
            Realm realm = null;
            try {
                realm = openRealm();
                realm.beginTransaction();
                RealmResults<Schedule> results = scheduleRepository.findAll(realm); //this is important
                Date yesterday = new Date();
                yesterday = DateUtils.addDays(MiDateUtils.setToEndOfDay(yesterday), -2);
                Log.i(TAG, "Starting to delete old schedules.");
                for(int i = 0; i < results.size(); i++) {
                    if(results.get(i).getEndTime().before(yesterday)) {
                        scheduleRepository.delete(realm, results.get(i--));
                    }
                }
                Log.i(TAG, "Deleting old schedules finishing...");
                realm.commitTransaction();
                Log.i(TAG, "Committed old schedule deletion transaction.");
            } finally {
                if(realm != null) {
                    realm.close();
                }
            }
        }

This took like 10 seconds to complete (and we execute this on start-up), this is executed on a background thread. But the UI froze up.

However, I replaced this with the following:

        @Override
        public void onRun()
                throws Throwable {
            Realm realm = null;
            try {
                realm = openRealm();
                realm.beginTransaction();
                RealmResults<Schedule> results = scheduleRepository.getSchedulesFromBeforeYesterday(realm);
                Log.i(TAG, "Starting to delete old schedules.");
                while(!results.isEmpty()) {
                    scheduleRepository.delete(realm, results.get(0));
                }
                Log.i(TAG, "Deleting old schedules finishing...");
                realm.commitTransaction();
                Log.i(TAG, "Committed old schedule deletion transaction.");
            } finally {
                if(realm != null) {
                    realm.close();
                }
            }
        }

And the ANRs stopped and is much much quicker.

    @Override
    public RealmResults<Schedule> getSchedulesFromBeforeYesterday(Realm realm) {
        Date yesterday = new Date();
        yesterday = DateUtils.addDays(MiDateUtils.setToEndOfDay(yesterday), -2);
        return query(realm).lessThan(Schedule.Fields.END_TIME.getField(), yesterday).findAll();
    }

The delete I'm calling here is just

    @Override
    public void delete(Realm realm, T t) {
        if(t != null) {
            t.removeFromRealm();
        }
    }

However, I still don't understand why there is a lock on the Realm of some sort that affects the UI thread if you are iterating a `RealmResults<T>` and deleting elements out of it.

**Comment by beeender** [comment]
@Zhuinden To delete all objects in a `RealmResults`, you can just call `RealmResults.clear()`

doc is here https://realm.io/docs/java/latest/api/io/realm/RealmResults.html#clear--

**Comment by Zhuinden** [comment]
@beeender well now I feel stupid, I'm going to use that because `nativeClear(nativePtr);` seems far more efficient than whatever I'm doing now

But that still doesn't really answer why deleting while iterating a large dataset can cause such a problem :P

**Comment by bmunkholm** [comment]
Write transactions are blocking calls, so if you try to do a single write in the UI tread you would be waiting for this to complete.

**Comment by Zhuinden** [comment]
@bmunkholm I see! That explains it then, so write transaction block across threads, and this needs to finish before that `second one` which would need to run. In our code, there's only one other start-up task that runs on start-up and actually attempts to make a write on the UI thread, so I guess these two are clashing against one another. I'll put that on the background thread and that should fix the problem.

Sorry about that! Apparently we were clearly at fault on this one - but at least we learned something new!

**Comment by dalinaum** [comment]
Hi @Zhuinden. I'll close this. If you encounter further problem, please feel free to reopen this or create a new issue.

**Comment by crysan** [comment]
I apologize for being late.
My problem was the following: base weighed about 118 Mb.  And sometimes there was a delay delete records.  After Realm.compactRealm (myConfig);  base decreased to 8 Mb.  After that, no delays when deleting records were observed.
I leave this post here in the hope that it is useful to someone.

**Comment by dalinaum** [comment]
@crysan Great. I had closed this issue. But it is a my mistake. I think it is created by @Zhuinden Sorry about that.

**Comment by dynamo-dma** [comment]
Hello guy. Please help me. i want to b delete one record in realm object. but not success.

 try {
            Realm realm = Realm.getInstance(getConfig());
            realm.beginTransaction();
            Person model =     realm.where(Person.class).equalTo("PersonID", "00001").findFirst();
            model.deleteFromRealm();
            realm.commitTransaction();
            realm.close();
        } catch (Exception e) {
            Log.i("On", "error:" + e.toString());
            e.printStackTrace();
        }

**Comment by Zhuinden** [comment]
Well based on what's here, it should work, so your error is elsewhere.
